File: ios/Packages/RideLinkCore/Sources/RideLinkCore/Model/Ulid.swift
import Foundation

/// Generates a fresh identifier in the 26-character Crockford-base32 shape `ManifestId`/
/// `TransferId` already require (both minted at the app layer: a `ManifestId` by whichever peer is
/// about to send `MANIFEST_BEGIN`, a `TransferId` by the requester in `TRANSFER_REQUEST`).
///
/// **Not** a spec-faithful monotonic ULID (a real ULID encodes a sortable 48-bit millisecond
/// timestamp; this is uniform CSPRNG output over the full 26 characters). Nothing here depends on
/// timestamp-sortability for a `ManifestId`/`TransferId` — only on the shape and on
/// unpredictability — mirroring `com.ridelink.core.model.Ulid`'s identical stance exactly.
///
/// `SystemRandomNumberGenerator` rather than `Security.framework`'s `SecRandomCopyBytes` (which
/// `BulkTokenTable` uses, for its own stronger token-secrecy requirement): that keeps this pure per
/// CLAUDE.md rule 9 (RideLinkCore imports only Foundation+CryptoKit), the same way Android's
/// mirror uses `java.security.SecureRandom` — the JVM standard library, not an Android platform
/// API — from inside its own pure `core` module.
public enum Ulid {
    private static let alphabet = Array("0123456789ABCDEFGHJKMNPQRSTVWXYZ")

    public static func generate() -> String {
        var rng = SystemRandomNumberGenerator()
        return String((0..<26).map { _ in alphabet[Int.random(in: 0..<alphabet.count, using: &rng)] })
    }
}

File: android/core/src/main/kotlin/com/ridelink/core/model/Ulid.kt
package com.ridelink.core.model

import java.security.SecureRandom

/**
 * Generates a fresh identifier in the 26-character Crockford-base32 shape [ManifestId]/
 * [TransferId] (and PROTOCOL's `session_id`/`msg_id`) already require.
 *
 * **Not** a spec-faithful monotonic ULID (real ULIDs encode a sortable 48-bit millisecond
 * timestamp; this is 130 bits of uniform CSPRNG output). Nothing in this codebase depends on
 * timestamp-sortability for a [ManifestId] or [TransferId] — only on the shape and on
 * unpredictability (brief §24: "unpredictable if used for authorization… bounded canonical
 * encoding"), both of which this satisfies. `network.control.ControlHandshake.freshSessionId`
 * already documents the same "shape now, real ULID semantics later if ever needed" stance for
 * `session_id`; this is that same low-risk follow-up, scoped to exactly the two new identifier
 * types that added a `require`d format.
 */
object Ulid {
    private const val ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
    private const val LENGTH = 26
    private val random = SecureRandom()

    fun generate(): String = CharArray(LENGTH) { ALPHABET[random.nextInt(ALPHABET.length)] }.concatToString()
}
